### ISSUE TITLE
ReadTheDocs fails with `invalid version: 'latest'`

### ISSUE THREAD

**Issue by jugmac00** [readthedocs/readthedocs.org#9805]
## Details

* Read the Docs project URL: https://readthedocs.org/projects/flask-reuploaded
* Build URL (if applicable): https://readthedocs.org/projects/flask-reuploaded/builds/18901216/
* Read the Docs username (if applicable): https://readthedocs.org/profiles/jugmac00

## Expected Result

Working build

## Actual Result

```
Running Sphinx v5.3.0
making output directory... done

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/sphinx/events.py", line 94, in emit
    results.append(listener.handler(self.app, *args))
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/pallets_sphinx_themes/theme_check.py", line 46, in wrapped
    return f(app, *args, **kwargs)
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/pallets_sphinx_themes/versions.py", line 14, in load_versions
    versions = readthedocs_versions(app)
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/pallets_sphinx_themes/versions.py", line 79, in readthedocs_versions
    DocVersion(name=name, slug=slug, dev=dev, current=slug == current_slug)
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/pallets_sphinx_themes/versions.py", line 116, in __new__
    version = pv.parse(slug)
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/packaging/version.py", line 52, in parse
    return Version(version)
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/packaging/version.py", line 197, in __init__
    raise InvalidVersion(f"Invalid version: '{version}'")
packaging.version.InvalidVersion: Invalid version: 'latest'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/sphinx/cmd/build.py", line 280, in build_main
    args.pdb)
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/sphinx/application.py", line 262, in __init__
    self._init_builder()
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/sphinx/application.py", line 335, in _init_builder
    self.events.emit('builder-inited')
  File "/home/docs/checkouts/readthedocs.org/user_builds/flask-reuploaded/envs/149/lib/python3.7/site-packages/sphinx/events.py", line 106, in emit
    (listener.handler, name), exc, modname=modname) from exc
sphinx.errors.ExtensionError: Handler <function load_versions at 0x7f7858c0c560> for event 'builder-inited' threw an exception (exception: Invalid version: 'latest')

Extension error (pallets_sphinx_themes.versions):
Handler <function load_versions at 0x7f7858c0c560> for event 'builder-inited' threw an exception (exception: Invalid version: 'latest')
```

This error occurred after updating my build dependencies...

```
diff --git a/docs/requirements.txt b/docs/requirements.txt
index 7b2883c..0ee299c 100644
--- a/docs/requirements.txt
+++ b/docs/requirements.txt
@@ -1,46 +1,44 @@
 #
-# This file is autogenerated by pip-compile with python 3.8
-# To update, run:
+# This file is autogenerated by pip-compile with Python 3.8
+# by the following command:
 #
 #    pip-compile --output-file=docs/requirements.txt docs/requirements.in
 #
 alabaster==0.7.12
     # via sphinx
-babel==2.9.1
+babel==2.11.0
     # via sphinx
-certifi==2021.10.8
+certifi==2022.12.7
     # via requests
-charset-normalizer==2.0.12
+charset-normalizer==2.1.1
     # via requests
-docutils==0.17.1
+docutils==0.19
     # via sphinx
-idna==3.3
+idna==3.4
     # via requests
-imagesize==1.3.0
+imagesize==1.4.1
     # via sphinx
-importlib-metadata==4.11.3
+importlib-metadata==5.1.0
     # via sphinx
-jinja2==3.1.1
+jinja2==3.1.2
     # via sphinx
 markupsafe==2.1.1
     # via jinja2
-packaging==21.3
+packaging==22.0
     # via
     #   pallets-sphinx-themes
     #   sphinx
 pallets-sphinx-themes==2.0.2
     # via -r docs/requirements.in
-pygments==2.11.2
+pygments==2.13.0
     # via sphinx
-pyparsing==3.0.8
-    # via packaging
-pytz==2022.1
+pytz==2022.6
     # via babel
-requests==2.27.1
+requests==2.28.1
     # via sphinx
 snowballstemmer==2.2.0
     # via sphinx
-sphinx==4.5.0
+sphinx==5.3.0
     # via
     #   -r docs/requirements.in
     #   pallets-sphinx-themes
@@ -56,7 +54,7 @@ sphinxcontrib-qthelp==1.0.3
     # via sphinx
 sphinxcontrib-serializinghtml==1.1.5
     # via sphinx
-urllib3==1.26.9
+urllib3==1.26.13
     # via requests
-zipp==3.8.0
+zipp==3.11.0
     # via importlib-metadata
```

So, `pallets-sphinx-themes` was not changed, but most notably the `packaging` version.

Building docs locally via tox works, and I am bit puzzled why this error is happening.

**Comment by jwag956** [comment]
Looks like the removal of 'LegacyVersion' is the issue - packaging uses to fall back and basically let any 'version' be created.(Version.parse() did a try/except - which has been removed).

**Comment by stsewd** [comment]
Hi, this doesn't look like a problem with Read the Docs, but with one of your dependencies, you should be able to replicate this issue locally.

**Comment by jugmac00** [comment]
> Hi, this doesn't look like a problem with Read the Docs, but with one of your dependencies, you should be able to replicate this issue locally.

No, it works locally.

```
❯ tox -r -e docs
GLOB sdist-make: /home/jugmac00/Projects/flask-reuploaded/setup.py
docs recreate: /home/jugmac00/Projects/flask-reuploaded/.tox/docs
docs installdeps: -rdocs/requirements.txt
docs inst: /home/jugmac00/Projects/flask-reuploaded/.tox/.tmp/package/1/Flask-Reuploaded-1.3.0.dev0.zip
docs installed: alabaster==0.7.12,attrs==22.1.0,Babel==2.11.0,certifi==2022.12.7,charset-normalizer==2.1.1,click==8.1.3,coverage==7.0.0,docutils==0.19,exceptiongroup==1.0.4,Flask==2.2.2,Flask-Reuploaded @ file:///home/jugmac00/Projects/flask-reuploaded/.tox/.tmp/package/1/Flask-Reuploaded-1.3.0.dev0.zip,idna==3.4,imagesize==1.4.1,importlib-metadata==5.1.0,iniconfig==1.1.1,itsdangerous==2.1.2,Jinja2==3.1.2,MarkupSafe==2.1.1,packaging==22.0,Pallets-Sphinx-Themes==2.0.2,pluggy==1.0.0,Pygments==2.13.0,pytest==7.2.0,pytest-cov==4.0.0,pytz==2022.6,requests==2.28.1,snowballstemmer==2.2.0,Sphinx==5.3.0,sphinxcontrib-applehelp==1.0.2,sphinxcontrib-devhelp==1.0.2,sphinxcontrib-htmlhelp==2.0.0,sphinxcontrib-jsmath==1.0.1,sphinxcontrib-qthelp==1.0.3,sphinxcontrib-serializinghtml==1.1.5,tomli==2.0.1,urllib3==1.26.13,Werkzeug==2.2.2,zipp==3.11.0
docs run-test-pre: PYTHONHASHSEED='4129648641'
docs run-test: commands[0] | sphinx-build -b html docs _build/html
Running Sphinx v5.3.0
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 9 source files that are out of date
updating environment: [config changed ('version')] 9 added, 0 changed, 0 removed
reading sources... [ 11%] api
reading sources... [ 22%] changes
reading sources... [ 33%] configuration
reading sources... [ 44%] contributing
reading sources... [ 55%] explanation
reading sources... [ 66%] getting_started
reading sources... [ 77%] index
reading sources... [ 88%] installation
reading sources... [100%] minimal_app

looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [ 11%] api
writing output... [ 22%] changes
writing output... [ 33%] configuration
writing output... [ 44%] contributing
writing output... [ 55%] explanation
writing output... [ 66%] getting_started
writing output... [ 77%] index
writing output... [ 88%] installation
writing output... [100%] minimal_app

generating indices... genindex done
writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded.

The HTML pages are in _build/html.
___________________________________ summary ____________________________________
  docs: commands succeeded
  congratulations :)
````

And thanks @jwag956 for having a look, too. Do you happen to know a way forward?  And what `versionǛ does this refer to? I have not defined "latest" anywhere (except of part of an URL in the codebase). That is also the reason why I think it is related to RTD.

**Comment by jwag956** [comment]
So currently - the way forward is to down-grade packages to 21.3

I believe the issue is mostly in pallets-sphinx-themes - which has different code path:
```
 if os.environ.get("READTHEDOCS"):
        versions = readthedocs_versions(app)
    else:
        versions = local_versions(app)
```

That code looks for versions:
`config_versions = app.config.html_context["versions"]`

And if you look at your RTD it likely has a 'version' == 'latest' and version == 'stable'. Those are being passed down via the above code to the packaging module and in the new world order - those are not valid semantic versions - and an exception is thrown. In packaging 21.x if the version didn't parse - it created a LegacyVersion which could be anything - and everything proceeded fine.

**Comment by jugmac00** [comment]
@jwag956 Thanks so much! I pinned `packaging < 22` for now and was able to build documentation again.

Do you think it makes sense to close this issue here and create one at the issue tracker for `pallets-sphinx-themes`?

**Comment by davidism** [comment]
That's a problem with Pallets-Sphinx-Theme, which I'm working on.

**Comment by jugmac00** [comment]
I created an issue over at https://github.com/pallets/pallets-sphinx-themes/issues/66 as the issue tracker there was also a place to look for known issues when I tried to debug the broken build, but did not find an open issue.

Thanks for you help, and @davidism thanks for working on a fix.

**Comment by davidism** [comment]
Fixed with Pallets-Sphinx-Themes 2.0.3.

**Comment by humitos** [comment]
It seems the problem is solved already. So, I'm closing this issue. Thanks you all for reporting and working on it 💯
